### PR TITLE
Added MD047: File should end with a blank line

### PR DESCRIPTION
**Pull request by androidseb**
Signed-off-by: androidseb <theandroidseb@gmail.com>

## Description
Adding a new rule ensuring the last line of a file is blank.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [X] Documentation (non-breaking change that does not add functionality but updates documentation)
- [ ] Chore (non-breaking change that does not add functionality or fix an issue)

## Checklist:
- [X] I have read the [**CONTRIBUTING**](https://github.com/markdownlint/markdownlint/blob/master/CONTRIBUTING.md) document.
- [X] Wrote [good commit messages](https://chris.beams.io/posts/git-commit/)
- [X] Feature branch is up-to-date with `master`, if not - rebase it
- [ ] Added tests for all new/changed functionality, including tests for positive and negative scenarios
- [X] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences

### CODE REVIEW

File: docs/RULES.md [hunk 1]
```diff
@@ -1087,3 +1088,28 @@ To fix this, used fenced code blocks:
...
 
 The reverse is true if the rule is configured to use the `indented` style.
+
+## MD047 - File should end with a blank line
+
+Tags: blank_lines
+
+Aliases: file-should-end-with-blank-line
+
+This rule is triggered if the last line of the file is not empty.
```
Comment (jaymzh) [inline docs/RULES.md]: That's not entirely accurate, it's more like "If the file does not end in a newline"

**Review by jaymzh — CHANGES_REQUESTED** [review]
Minor tweaks, but this seems good. However, we try to keep rule numbers in sync with @DavidAnson 's project... so lets check in with him on that.

**Comment by DavidAnson** [comment]
@jaymzh Thanks for checking! MD047 is, indeed, already used but it appears to be for this exact purpose! I don't know if that was chance or luck, but the planets aligned well today.

https://github.com/DavidAnson/markdownlint/blob/main/doc/Rules.md#md047---files-should-end-with-a-single-newline-character

**Comment by androidseb** [comment]
Oh wow! That is pretty good, I didn't even do it on purpose, the documentation of the rule is so similar too!
@DavidAnson's wording for the doc and the rule alias looks better, maybe I can switch to this, this will help with consistency.
What do you think @jaymzh?

**Comment by jaymzh** [comment]
@androidseb - yup, sounds like a good idea.

Also you have some failing tests :)

**Comment by androidseb** [comment]
@jaymzh I have looked at the failing tests and it seems a lot of tests are failing because the input sample data doesn't end with an empty line.
Question 1:
What would you recommend in this case?
I could edit the input data everywhere in the test texts to end with "\n", would that be the proper way to do this?

**Comment by androidseb** [comment]
Question 2: I have added some tests too, could you take a look and let me know if they look good to you please?

File: test/test_cli.rb [hunk 2]
```diff
@@ -224,6 +224,25 @@ def test_ignore_front_matter
     assert_equal(result[:stdout], expected)
   end
 
+  def test_file_ends_with_single_newline_character_good
```
Comment (jaymzh) [inline test/test_cli.rb]: this file is meant to test command-line options, not rules. Can you instead add a test to `tests/rule_tests`? It's really easy you just create a file and add `{MD047}` on the line that should trigger.

**Comment by jaymzh** [comment]
ping @androidseb

**Comment by androidseb** [comment]
Hey, sorry I lost momentum on this one, but I want to get back at this at one point. Is this blocking anything on your side?

**Comment by jaymzh** [comment]
Well, I wanted this feature the other day :) but other than that, no

**Comment by androidseb** [comment]
@jaymzh OK, I'm getting back to this, this tab has been open in my browser for months :sweat_smile:

So after adding some tests for the rule, I noticed some other tests are failing in `test_cli.rb` and it's most likely because some of those tests use sample MD content not ending with ~an empty line~ a single newline character.

How should I fix these?

1. Change the input text to end with a line return?
2. Change the "expected/actual" check of the test?

**Comment by jaymzh** [comment]
Oh man I completely misread this. I thought you wanted the file to end-in-a-newline, as opposed to end in a blankline (two newlines).

That's... the opposite of most linters, almost any linter I've seen will complain about a blank line at the end. I'm happy to have a rule for anything but we should make the default not require a blank line.

That will also answer your test_cli question, which is don't add arbitrary blank lines to the tests please.

**Comment by androidseb** [comment]
Sorry if mis-communicated, but no, I'm not trying to get 2 newline characters, only a single one, as the name of the rule indicates: `file_ends_with_single_newline_character`.

Maybe I misunderstood the error from the test logs. Here is what I can see in the test logs:
```
 26) Failure:
TestCli#test_json_output [/home/runner/work/markdownlint/markdownlint/test/test_cli.rb:16]:
Expected: 0
  Actual: 1

266 runs, 356 assertions, 26 failures, 0 errors, 0 skips
rake aborted!
Command failed with status (1)
```

From my understanding, the problem comes from here:
```ruby
  def test_json_output
    result = run_cli_with_input('-j', '# header')
    assert_ran_ok(result)
    assert_equal("[]\n", result[:stdout])
  end
```

This looks like the content of the MD text being tested here will be:
```
# header<EOF>
```
instead of
```
# header
<EOF>
```

To fix this, one way to do it would be writing the test code like this:
```ruby
  def test_json_output
    result = run_cli_with_input('-j', "# header\n")
    assert_ran_ok(result)
    assert_equal("[]\n", result[:stdout])
  end
```

I edited my previous comment to make my message clearer.

File: test/rule_tests/file_ends_with_single_newline_character_good copy.md [hunk 3]
```diff
@@ -0,0 +1,3 @@
+# File ending with an empty line
```
Comment (jaymzh) [inline test/rule_tests/file_ends_with_single_newline_character_good copy.md]: this probably shouldn't be called ` copy` ?

**Review by jaymzh — COMMENTED** [review]
Ah I see. So yeah we should fix those tests to pass in a newline, i.e.:

```ruby
run_cli_with_input('-j', "# header\n")
```

That's clearly more realistic input.

What threw me off was some of your other test cases, which when I was viewing on mobile seemed to show two newlines, but now that I check on a real system all look good.

So yeah, fix up the test_cli cases and I think we're good to go!

**Comment by jaymzh** [comment]
code looks good, still got got some test failures though

**Comment by androidseb** [comment]
@jaymzh Please check it out, tests are passing now, but I had to edit quite a few test files, let me know if that looks all good to you, thanks!

**Review by jaymzh — APPROVED** [review]
Looks great, thanks!

This is going to be complete merge hell for my PR on lists... but that's OK. :)

Thanks!

**Comment by jaymzh** [comment]
Hey @androidseb how did you make files *without* a newline on the end? I can't seem to do it in vim, and after I rebased my PR on this I can't fix `lists_without_blank_lines.md` which vim helpfully "fixes" for me.

**Comment by jaymzh** [comment]
Eh nevermind, stackoverflow to the rescue.

For the record:
```
vim -b ...
:set noeol
:wq
```

**Comment by androidseb** [comment]
@jaymzh I used [Visual Studio Code](https://code.visualstudio.com/) to make those edits, glad you figured it out with vim :+1:

**Pull request merged**